File: src/main/js/front/app/controller/InboxController.js
import { API_URL } from '../constants';

// Function to get the house invitations of a user
export const getUsersHouseInvitations = async (userId) => {
    try {
        const response = await fetch(`${API_URL}/getInbox/${userId}`, {
            method: 'GET',
            headers: {
                'Content-Type': 'application/json',
            },
        });
        if (!response.ok) {
            const errorMessage = await response.text();
            throw new Error(errorMessage);
        }
        return await response.json(); // This will be the list of house IDs
    } catch (error) {
        console.error("Failed to get inbox:", error);
        throw error;
    }
};

// Function to get the near expirations products of the house of a user
export const getNearExpirationStocks = async (userId) => {
    try {
        const response = await fetch(`${API_URL}/getSoonToExpire/${userId}`, {
            method: 'GET',
            headers: {
                'Content-Type': 'application/json',
            },
        });
        if (!response.ok) {
            const errorMessage = await response.text();
            throw new Error(errorMessage);
        }
        return await response.json(); // This will be the list of products
    } catch (error) {
        console.error("Failed to get near expirations:", error);
        throw error;
    }
};

// Function to receive chat notifications of a user
export const getChatNotifications = async (userId) => {
    try {
        const response = await fetch(`${API_URL}/getChatNotifications/${userId}`, {
            method: 'GET',
            headers: {
                'Content-Type': 'application/json',
            },
        });
        if (!response.ok) {
            const errorMessage = await response.text();
            throw new Error(errorMessage);
        }
        return await response.json(); // This will be the list of chat notifications
    } catch (error) {
        console.error("Failed to get chat notifications:", error);
        throw error;
    }
};

// Function to get the inbox size of a user
export const getInboxSize = async (userId) => {
    try {
        const response = await fetch(`${API_URL}/getInboxSize/${userId}`, {
            method: 'GET',
            headers: {
                'Content-Type': 'application/json',
            },
        });
        if (!response.ok) {
            const errorMessage = await response.text();
            throw new Error(errorMessage);
        }
        return await response.json(); // This will be the inbox size
    } catch (error) {
        console.error("Failed to get inbox size:", error);
        throw error;
    }
};

File: src/main/java/org/austral/ing/lab1/controller/InboxController.java
package org.austral.ing.lab1.controller;

import com.google.gson.Gson;
import org.austral.ing.lab1.object.jsonparsable.ExpirationInfo;
import org.austral.ing.lab1.repository.chats.Chats;
import org.austral.ing.lab1.repository.inboxes.ChatNotifications;
import org.austral.ing.lab1.repository.inboxes.HouseInvitations;
import org.austral.ing.lab1.repository.inboxes.NearExpirations;
import org.austral.ing.lab1.repository.inboxes.Notifications;
import spark.Spark;

import javax.persistence.EntityManager;
import javax.persistence.EntityManagerFactory;
import java.util.List;
import java.util.Map;

public class InboxController {
    private final Gson gson = new Gson();
    private final EntityManagerFactory entityManagerFactory;

    public InboxController(EntityManagerFactory entityManagerFactory) {
        this.entityManagerFactory = entityManagerFactory;
    }

    public void init() {

        // Route to get the house invitations of a given user
        Spark.get("/inbox/getHouseInvitations", (req, resp) -> {
            EntityManager entityManager = entityManagerFactory.createEntityManager();
            HouseInvitations houseInvitations = new HouseInvitations(entityManager);
            try {
                Long userId = Long.parseLong(req.headers("UserId"));
                List<Map<String, Object>> inboxMessage = houseInvitations.getHousesByUserId(userId);
                resp.status(200);
                resp.type("application/json");
                return gson.toJson(inboxMessage);
            } catch (Exception e) {
                resp.status(500);
                System.out.println(e.getMessage());
                return "An error occurred while getting the inbox, please try again";
            } finally {
                entityManager.close();
            }
        });

        // Route to get the soon to expire products of a given user
        Spark.get("/inbox/getSoonToExpire", (req, resp) -> {
            EntityManager entityManager = entityManagerFactory.createEntityManager();
            NearExpirations expirationsRepo = new NearExpirations(entityManager);
            try {
                Long userId = Long.parseLong(req.headers("UserId"));
                List<ExpirationInfo> inboxMessage = expirationsRepo.getExpiringStocks(userId);
                resp.status(200);
                resp.type("application/json");
                return gson.toJson(inboxMessage);
            } catch (Exception e) {
                resp.status(500);
                System.out.println(e.getMessage());
                return "An error occurred while getting the soon to expire products, please try again";
            } finally {
                entityManager.close();
            }
        });

        // Route to get the chat notifications of a user
        Spark.get("/inbox/getChatNotifications", (req, resp) -> {
            EntityManager entityManager = entityManagerFactory.createEntityManager();
            Chats chatNotifications = new Chats(entityManager);
            try {
                Long userId = Long.parseLong(req.headers("UserId"));
                String chatsNotifications = gson.toJson(chatNotifications.getNotifications(userId));
                resp.status(200);
                resp.type("application/json");
                return chatsNotifications;
            } catch (Exception e) {
                resp.status(500);
                System.out.println(e.getMessage());
                return "An error occurred while getting the chat notifications, please try again";
            } finally {
                entityManager.close();
            }
        });

        // Route to get the sizes of the different notifications of a user (unused)
        Spark.get("/inbox/getInboxSize", (req, resp) -> {
            EntityManager entityManager = entityManagerFactory.createEntityManager();
            Notifications notifications = new Notifications(entityManager);
            try {
                Long userId = Long.parseLong(req.headers("UserId"));
                Map<String, Long> notificationCounts = notifications.getNotificationCounts(userId);
                resp.status(200);
                resp.type("application/json");
                return gson.toJson(notificationCounts);
            } catch (Exception e) {
                resp.status(500);
                System.out.println(e.getMessage());
                return "An error occurred while getting the notification sizes, please try again";
            } finally {
                entityManager.close();
            }
        });
    }
}
